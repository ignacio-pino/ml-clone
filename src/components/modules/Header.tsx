import { SearchIcon } from "@chakra-ui/icons";
import {
  Flex,
  Input,
  Image,
  InputGroup,
  InputRightElement,
  Checkbox,
  IconButton,
  Box,
  Divider,
  HStack,
} from "@chakra-ui/react";
import { useState } from "react";
import { createSearchParams, useLocation, useNavigate } from "react-router-dom";
import { IHeaderCategoryData } from "../../Interfaces";

export default function Header({
  categoryId,
  categoryName,
}: IHeaderCategoryData) {
  const { search } = useLocation();
  const searchParams = new URLSearchParams(search);

  const navigate = useNavigate();

  let params = Object.fromEntries(searchParams.entries());

  const [searchValue, setSearchValue] = useState("");
  const [categoryCheck, setCategoryCheck] = useState(false);

  const handleClick = (
    event:
      | React.MouseEvent<HTMLButtonElement, MouseEvent>
      | React.FormEvent<HTMLFormElement>
  ) => {
    event.preventDefault();
    searchValue
      ? categoryCheck
        ? (params = {
            q: searchValue.trim().toLocaleLowerCase(),
            category: categoryId,
          })
        : (params = { q: searchValue.trim().toLocaleLowerCase() })
      : delete params.q;
    navigate({
      pathname: "/products",
      search: `${createSearchParams(params)}`,
    });
  };

  return (
    <Flex
      w="100vw"
      h="10vh"
      backgroundColor="gray.900"
      boxShadow="lg"
      justifyContent="center"
      alignItems="center"
    >
      <Flex justifyContent="space-evenly" alignItems="center">
        <Image
          src="/assets/handshake.png"
          width="100px"
          height="100px"
          mr="2rem"
        />
        <InputGroup>
          <form onSubmit={(event) => handleClick(event)}>
            <Input
              variant="filled"
              placeholder="Busca productos, marcas y mÃ¡s..."
              w="3xl"
              value={searchValue}
              onChange={(event) => setSearchValue(event.currentTarget.value)}
            />
          </form>
          <InputRightElement>
            <Box
              display="flex"
              alignItems="center"
              justifyContent="flex-end"
              width="100%"
              gap={3}
            >
              {categoryName && (
                <HStack gap={5}>
                  <Divider orientation="vertical" color="gray.500" h="20px" />

                  <Checkbox
                    whiteSpace="nowrap"
                    color="gray.500"
                    onChange={() => setCategoryCheck(!categoryCheck)}
                    isChecked={categoryCheck}
                  >
                    Solo en {categoryName}
                  </Checkbox>

                  <Divider orientation="vertical" color="gray.500" h="20px" />
                </HStack>
              )}
              <IconButton
                aria-label="Buscar"
                icon={<SearchIcon />}
                onClick={(event) => handleClick(event)}
              />
            </Box>
          </InputRightElement>
        </InputGroup>
      </Flex>
    </Flex>
  );
}
