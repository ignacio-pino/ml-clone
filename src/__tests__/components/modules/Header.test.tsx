import { render, screen, fireEvent } from "@testing-library/react";
import Header from "../../../components/modules/Header";
import { MemoryRouter } from "react-router-dom";

const mockedUseNavigate = vitest.fn();

vitest.mock("react-router-dom", () => ({
  ...require("react-router-dom"),
  useNavigate: () => mockedUseNavigate,
}));

describe("Header Component", () => {
  const mockHeaderData = {
    categoryId: "categoryIdTest",
    categoryName: "categoryNameTest",
  };

  it("renders the search input correctly", () => {
    render(
      <MemoryRouter>
        <Header {...mockHeaderData} />
      </MemoryRouter>
    );
    const input = screen.getByPlaceholderText(
      "Busca productos, marcas y más..."
    );
    expect(input).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const container = render(
      <MemoryRouter>
        <Header {...mockHeaderData} />
      </MemoryRouter>
    ).container;

    expect(container).toMatchSnapshot();
  });

  it("updates search value when typing into the search input", () => {
    render(
      <MemoryRouter>
        <Header {...mockHeaderData} />
      </MemoryRouter>
    );
    const input = screen.getByPlaceholderText(
      "Busca productos, marcas y más..."
    );
    fireEvent.change(input, { target: { value: "test search" } });
    expect(input).toHaveValue("test search");
  });

  it("checks the category checkbox when clicked", () => {
    render(
      <MemoryRouter>
        <Header {...mockHeaderData} />
      </MemoryRouter>
    );
    const checkbox = screen.getByRole("checkbox", {
      name: `Solo en ${mockHeaderData.categoryName}`,
    });
    fireEvent.click(checkbox);
    expect(checkbox).toBeChecked();
  });

  it("calls navigate with correct parameters on search button click", () => {
    render(
      <MemoryRouter>
        <Header {...mockHeaderData} />
      </MemoryRouter>
    );
    const input = screen.getByPlaceholderText(
      "Busca productos, marcas y más..."
    );
    const checkbox = screen.getByRole("checkbox", {
      name: `Solo en ${mockHeaderData.categoryName}`,
    });
    const searchButton = screen.getByLabelText("Buscar");

    fireEvent.change(input, { target: { value: "test search" } });
    fireEvent.click(checkbox);
    fireEvent.click(searchButton);

    expect(mockedUseNavigate).toHaveBeenCalledWith({
      pathname: "/products",
      search: "q=test+search&category=categoryIdTest",
    });
  });

  it("navigates with both searchValue and category when checkbox is checked", () => {
    render(
      <MemoryRouter>
        <Header {...mockHeaderData} />
      </MemoryRouter>
    );

    const input = screen.getByPlaceholderText(
      "Busca productos, marcas y más..."
    );
    const checkbox = screen.getByRole("checkbox", {
      name: `Solo en ${mockHeaderData.categoryName}`,
    });
    fireEvent.change(input, { target: { value: "test search" } });
    fireEvent.click(checkbox);
    fireEvent.click(screen.getByLabelText("Buscar"));

    expect(mockedUseNavigate).toHaveBeenCalledWith({
      pathname: "/products",
      search: "q=test+search&category=categoryIdTest",
    });
  });

  it("navigates with only searchValue when checkbox is not checked", () => {
    render(
      <MemoryRouter>
        <Header {...mockHeaderData} />
      </MemoryRouter>
    );

    const input = screen.getByPlaceholderText(
      "Busca productos, marcas y más..."
    );
    fireEvent.change(input, { target: { value: "test search" } });
    fireEvent.click(screen.getByLabelText("Buscar"));

    expect(mockedUseNavigate).toHaveBeenCalledWith({
      pathname: "/products",
      search: "q=test+search",
    });
  });

  it("navigates without a searchValue when input is empty", () => {
    render(
      <MemoryRouter>
        <Header {...mockHeaderData} />
      </MemoryRouter>
    );

    fireEvent.click(screen.getByLabelText("Buscar"));

    expect(mockedUseNavigate).toHaveBeenCalledWith({
      pathname: "/products",
      search: "",
    });
  });

  it("calls handleClick function on submit", () => {
    render(
      <MemoryRouter>
        <Header {...mockHeaderData} />
      </MemoryRouter>
    );

    const input = screen.getByPlaceholderText(
      "Busca productos, marcas y más..."
    );
    fireEvent.change(input, { target: { value: "test search" } });
    fireEvent.submit(input);

    expect(mockedUseNavigate).toHaveBeenCalledWith({
      pathname: "/products",
      search: "q=test+search",
    });
  });
});
