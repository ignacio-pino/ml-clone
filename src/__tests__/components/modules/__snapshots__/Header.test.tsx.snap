// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header Component > matches snapshot 1`] = `
<div>
  <div
    class="css-1qgk9u0"
  >
    <div
      class="css-1vr8h81"
    >
      <img
        class="chakra-image css-cbs51e"
        src="/assets/handshake.png"
      />
      <div
        class="chakra-input__group css-bx0blc"
        data-group="true"
      >
        <form>
          <input
            class="chakra-input css-1ybc3d8"
            placeholder="Busca productos, marcas y mÃ¡s..."
            value=""
          />
        </form>
        <div
          class="chakra-input__right-element css-dp9r5f"
        >
          <div
            class="css-1c2c1lm"
          >
            <div
              class="chakra-stack css-cp3a5l"
            >
              <hr
                aria-orientation="vertical"
                class="chakra-divider css-lcs0ep"
              />
              <label
                class="chakra-checkbox css-1v4t15a"
              >
                <input
                  class="chakra-checkbox__input"
                  style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                  type="checkbox"
                  value=""
                />
                <span
                  aria-hidden="true"
                  class="chakra-checkbox__control css-dnty2r"
                />
                <span
                  class="chakra-checkbox__label css-1oeb2oe"
                >
                  Solo en 
                  categoryNameTest
                </span>
              </label>
              <hr
                aria-orientation="vertical"
                class="chakra-divider css-lcs0ep"
              />
            </div>
            <button
              aria-label="Buscar"
              class="chakra-button css-p7mkev"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="chakra-icon css-onkibi"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M23.384,21.619,16.855,15.09a9.284,9.284,0,1,0-1.768,1.768l6.529,6.529a1.266,1.266,0,0,0,1.768,0A1.251,1.251,0,0,0,23.384,21.619ZM2.75,9.5a6.75,6.75,0,1,1,6.75,6.75A6.758,6.758,0,0,1,2.75,9.5Z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
